src/dmd/root/string.d

```diff
@@ -212,3 +212,82 @@ unittest
     assert(dstrcmp("안녕하세요!", "안녕하세요!") == 0);
     static assert(dstrcmp("안녕하세요!", "안녕하세요!") == 0);
 }
+
+/**
+ * Infers the length `N` of a string literal and coerces its type to a static
+ * array with length `N + 1`. Returns the string with a null character appended
+ * to the end.
+ *
+ * Params:
+ *  literal = string literal
+ *
+ * Notes:
+ *  - LDC produces quite optimal code for short strings:
+ *    https://d.godbolt.org/z/mffrjy,
+ *    https://gist.github.com/PetarKirov/338e4ab9292b6b2b311a3070572a07fb (backup URL)
+*/
+T[N + 1] toStaticArray(T, size_t N)(scope const(T)[N] literal)
+if (is(T == char) || is(T == wchar) || is(T == dchar))
```

Comment:
It's never anything but `char` so no need for `T` parameter.

Reply:
I plan to use it in the future PRs, where I expect `wchar` and `dchar` support to come handy, but if you insist I can remove it for now.

BTW, in the beginning I was considering making this fully generic, like https://dlang.org/phobos/std_array#staticArray, but since the idea here to add null character at the end, I decided it's not worth the effort.

Reply:
Since this is specific to dmd, there's no need to make it generic in ways that are not useful to dmd.

Reply:
> Since this is specific to dmd, there's no need to make it generic in ways that are not useful to dmd.

Yes, that's why I allowed only `wchar` and `dchar`, in addition to `char`.

Reply:
which are not used in dmd

Reply:
From the top of my head, I've seen them used in `dmd.root.filename` and `dmd.utf`. `dmd.link` and `dmd.vsoptions` need to deal with the Windows API (which use wchar), so it may be handy there as well.

Though I guess where we disagree, is that I believe that a function should be made as generic as possible (within reasonable limits). Putting artificial limitations is a code smell and accomplishes nothing.

Reply:
non-char should only be used in DMD as the final step to send to/from a Windows API. Processing should be done as char otherwise.

> I believe that a function should be made as generic as possible

This is not a generic library. Building in unused features leads to bloat, bugs from untested code, and wastes peoples' time wondering why it is there.

I checked dmd.root.filename. It does indeed use wchar, but only to send to Windows APIs. Not process them. It never uses dchar.